style(all): fixing pipelines

ollibowers opened the pull request:
- created 7 new jobs out of the previous 2.
  - seperate lint, typecheck and test jobs for backend and frontend
  - one docker build and start prod job that runs if all others pass
  - (currently test jobs are disabled as these fixes will come separately)
- fixed pylint errors
  - for some, i added explicit ignores as i felt they were appropriately broken
  - for some, i added explicit ignores with todos, as i do not want to refactor completely until we get tests
  - removed some redundant or poorly placed pylint disables
  - intentionally left docstring disables in there as we will fix these in a future PR
  - similarly, intentionally left circular import ignores as i will fix this next in a utility refactor + user dict removal PR
  - **reworked the opening to run_app.py**, so its less annoying
- fixed eslint disables
  - removed a bunch of bad eslint disables, only leaving the necessary ones in
  - disabled no-console and allowed non-default exports for the api directory
  - **refactored term planner** decently, as the eslint-disable removal required it. this was a much needed refactor and is one step further to getting TermPlanner.tsx readable. 
  - moved some axios requests into the api folder, so i could remove some other eslint-disables
    - all will be moved in a upcoming PR (currently assigned to a subbie)
  - **reworked graphical selector** slightly to rely more on the unlocked courses route, as previously it was trying to use a `.unlocked` property on the user courses response (which didn't exist). now we have colors again!!!
- verified that there are no mypy `type: ignore`s that we do not want (or that wont get culled by something else very soon)
- verified that there are no tsc `@ts-nocheck`s that we do not want
- update team.md

ollibowers committed:
wip: re enable pylint for 331-staging

ollibowers committed:
wip: trying to disable the frontend tests and adding linting and type checks

ollibowers committed:
fix: forgor working directory

ollibowers committed:
fix: install dotenv for frontend

ollibowers committed:
wip: add workplace extensions and LF the .gitignore

ollibowers committed:
wip: testing out a composite action

ollibowers committed:
fix: forgor shell from my composite action

ollibowers committed:
wip: fix for pip install dotenv fail

ollibowers committed:
wip: i might be dumb

ollibowers committed:
wip: more testing, double python usage

ollibowers committed:
wip: added eslint job to test if it all works

ollibowers committed:
fix: wrong lts version

ollibowers committed:
wip: added first few jobs

ollibowers committed:
wip: added the backend jobs

ollibowers committed:
fix: backend tests failed, added wait for backend container

ollibowers committed:
chore: removed the old pylinter file

ollibowers committed:
fix: remove shell from sleep

ollibowers committed:
fix: pylint - trailing-whitespace, trailing-newlines, missing-final-newline

ollibowers committed:
fix: pylint - ungrouped-imports, wrong-import-position, reimproted, wrong-import-order

Also changed run_app.py checks to no longer require nodemon.

ollibowers committed:
fix: style - too-many-locals, too-many-function-args, cell-var-from-loop

Mostly just added scope pylint disables because I am not refactoring this code.

ollibowers committed:
fix: pylint - possibly-used-before-assignment, unused-import

ollibowers committed:
fix: pylint - unused-variable

ollibowers committed:
fix: pylint - multiple-statements, redefined-outer-name, no-else-return, consider-using-enumerate

ollibowers committed:
fix: pylint - bare-except, use-dict-literal

ollibowers committed:
fix: actual troll

ollibowers committed:
fix: pylint - subprocess-run-check, consider-using-sys-exit

ollibowers committed:
fix: pylint - consider-using-with, abstract-method, inconsistent-return-statements, unspecified-encoding

ollibowers committed:
wip: disable the unit tests for now

Peedee2002 commented on .github/workflows/ci.yml:
```diff
@@ -14,78 +14,112 @@ jobs:
         run: docker compose up -d ci-backend
-      - name: Run mypy checks
-        run: cd backend && python -m mypy . && cd ..
+      - name: Wait for backend to initialise  # cringe as hell, remove when we get health checks to wait on
```
Yea sorry oops

I didn't know about health checks when I wrote that cringe

ollibowers replied:
nonononnooononononon, not calling previous writers cringe, you are not cringe 🥺. calling my sleep timeout a cringe workaround

ollibowers committed:
fix: pylint - too-many-nested-blocks, too-many-return-statements, too-many-branches, too-many-public-methods

Most of this was just disables, as I believe these were used fairly and solemnly.

ollibowers committed:
fix: pylint - empty-docstring (and some new todo comments)

ollibowers committed:
feat: added healthcheck for backend containers to be used in ci, created a /ping route to achieve this

ollibowers committed:
feat: docker build job now requires all other jobs, as pointless check otherwise

ollibowers committed:
fix: re-disabled the backend tests

ollibowers committed:
chore: add 2024 team.md and update previous years for missing members

ollibowers committed:
chore: add 2022 luke github username

ollibowers committed:
fix: remove test jobs from docker job needs

ollibowers committed:
fix: pylint - remove redundant and leftover pylint disables

ollibowers committed:
fix: eslint - allow console.error and non-default exports in the api directory (until error reporting is standardised)

ollibowers committed:
wip: eslint - removed any redundant eslint-disabled

ollibowers committed:
chore: added some todo comments for backend type removal

ollibowers committed:
Merge branch 'style/all/fix-pylint-and-eslint' of github.com:csesoc/Circles into style/all/fix-pylint-and-eslint

ollibowers committed:
fix: eslint - removed the eslint-disable from TermPlanner and cleaned it up a lot

ollibowers committed:
fix: changed courseQueries.isPending to be isPlaceholderData since pending is always false there

ollibowers committed:
fix: eslint - removed a bunch of unnecessary eslint disables

ollibowers committed:
fix: eslint - moved out two axios calls as to resolve some eslint issues in their users

ollibowers committed:
fix: refactored graph to use unlocked course state for coloring as .unlocked doesnt exist on user courses

sachk reviewed (approved):
not super thorough but lgtm, if you want more feedback u can ask jared since he's more experienced with gh actions.

martanman reviewed (approved):
lgtm
